drgn/helpers/common/stack.py

```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""
+Stack
+-----
+
+The ``drgn.helpers.common.stack`` module provides helpers for working with the
+stack trace.
+"""
+
+from typing import List
+
+from drgn import FaultError, PlatformFlags, Program, StackFrame, StackTrace
+from drgn.helpers.common.memory import identify_address
+
+__all__ = ("print_annotated_stack",)
+
+
+# Split stack trace in to groups of stack frames
+# delineated by interrupted frame:
+# [[frame1, frame2, ...], [interrupted_frame, next frame, ...], ...]
+def _split_stack_trace(trace: StackTrace) -> List[List[StackFrame]]:
+    cur_group: List[StackFrame]
+    cur_group = []
+    groups = []
+
+    for frame in trace:
+        if frame.interrupted:
+            groups.append(cur_group)
+            cur_group = [frame]
+        else:
+            cur_group.append(frame)
+
+    groups.append(cur_group)  # final group
+    return groups
+
+
+def _print_annotated_frames_group(frames: List[StackFrame], prog: Program) -> None:
+    # platform must be known if there is a stack trace
+    assert prog.platform is not None
+    word_size = 8 if prog.platform.flags & PlatformFlags.IS_64_BIT else 4
+
+    frames_addrs = [frame.register("rsp") for frame in frames]
+    hex_format = "{:016x}" if word_size == 8 else "{:08x}"
+
+    start_addr = frames_addrs[0]
+    end_addr = frames_addrs[-1] + word_size - 1
+    stack_size = end_addr - start_addr + 1
+
+    try:
+        stack_bytes = prog.read(start_addr, stack_size)
+    except FaultError:
+        # addresses not mapped
+        # for e.g, frames in userspace
+        for frame in frames:
+            print(f"[stack frame {frame}]")
+
+        return
+
+    frame_ind = 0
+
+    for offset in range(0, len(stack_bytes), word_size):
+        addr = start_addr + offset
+        hex_addr = hex_format.format(addr)
+
+        word_bytes = stack_bytes[offset : offset + word_size]
+
+        if prog.platform.flags & PlatformFlags.IS_LITTLE_ENDIAN:
+            word_val = int.from_bytes(word_bytes, "little")
+        else:
+            word_val = int.from_bytes(word_bytes, "big")
+
+        # multiple frames can have the same address
+        # for e.g, the first context_switch frame
+        while frame_ind < len(frames_addrs) and addr == frames_addrs[frame_ind]:
+            frame = frames[frame_ind]
+            print(f"[stack frame {frame}]")
+
+            frame_ind += 1
+        else:
+            formatted_value = hex_format.format(word_val)
+
+            identified = identify_address(prog, word_val)
+            if identified:
+                formatted_value += f" [{identified}]"
+
+            print(f"{hex_addr}: {formatted_value}")
+
+
+def print_annotated_stack(trace: StackTrace) -> None:
+    """
+    Print the content of the stack trace, identifying symbols and
+    slab objects.
+
+    :param trace: Stack trace to print.
+    """
+    prog = trace.prog
+
+    print("STACK POINTER     VALUE")
```

Comment:
Make sure that this is still aligned nicely on 32-bit, since those stack pointers are only 8 characters wide. Maybe we can just pad the stack pointer with extra spaces on 32-bit.

Reply:
Hmmm I added some logic to handle this:
```
    if is_64_bit:
        print("STACK POINTER     VALUE")
    else:  # 8 extra spaces
        print("STACK POINTER             VALUE")
```
We have to figure out how to test it on a 32-bit platform tho 🤔

Reply:
Hardcoding a bunch of things to test the alignment:
<img width="376" alt="Screen Shot 2022-11-07 at 2 00 17 PM" src="https://user-images.githubusercontent.com/21051676/200424674-9745295a-0033-480b-a364-1ab920098e5b.png">

Reply:
(Basically just added 8 spaces)
```
                if is_64_bit:
                    print(f"{hex_addr}: {formatted_value}")
                else:
                    # 8 extra spaces to align the value
                    print(f"{hex_addr}:         {formatted_value}")
```